src/Lean/Elab/Do.lean

```diff
@@ -693,11 +693,11 @@ private def expandDoIf? (stx : Syntax) : MacroM (Option Syntax) := match stx wit
     let mut eIsSeq := true
     for (i, cond, t) in Array.zip (is.reverse.push i) (Array.zip (conds.reverse.push cond) (ts.reverse.push t)) do
       e ← if eIsSeq then pure e else `(doSeq|$e:doElem)
-      e ← withRef cond <| match cond with
-        | `(doIfCond|let $pat := $d) => `(doElem| match%$i $d:term with | $pat:term => $t | _ => $e)
-        | `(doIfCond|let $pat ← $d)  => `(doElem| match%$i ← $d    with | $pat:term => $t | _ => $e)
-        | `(doIfCond|$cond:doIfProp) => `(doElem| if%$i $cond:doIfProp then $t else $e)
-        | _                          => `(doElem| if%$i $(Syntax.missing) then $t else $e)
+      e ← withRef (mkNullNode #[i, t]) <| match cond with
```

Comment:
This could still produce a negative range if `i` and `t` were swapped, or am I missing something?

Reply:
Yes, that's true. It looks like this really was working around a position loss in `if`, so I fixed that one instead

Reply:
We do have a few more occurrences of this pattern, but there the chance of getting swapped is much lower than here I'd say.